[HTTPS] Adds PEM support for Kestrel

javiercn opened the pull request:
* Allows loading protected and unprotected PEM key files combined with a DER (crt) certificate

### Open questions.
- [x] Is there a better way to detect the key type other than probing?
- [x] ECDsa support.

Addresses #4706

javiercn committed:
Add additional certificate formats

javiercn committed:
Add PEM support

javiercn committed:
Force the keys to be persisted on windows

javiercn committed:
Add an E2E test for ensuring the key works

Tratcher commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+
+            static X509Certificate2 PersistKey(X509Certificate2 fullCertificate)
+            {
+                // We need to force the key to be persisted.
```
Persisted to where?

blowdart replied:
On Windows this needs to be super careful, otherwise you can end up filling the private key directory. @bartonjs will know more

bartonjs replied:
This is actually doing what I call a "perphemeral" key, in that it's "temporarily persisted", as long as the cert gets freed (Dispose or Finalize) the key-on-disk gets erased.

Unfortunately, SChannel doesn't work without a key-on-disk (or key-in-HSM), so this _is_ required after loading the private key from a PEM file.

JamesNK replied:
> as long as the cert gets freed (Dispose or Finalize) the key-on-disk gets erased.

aspnet people, is the certificate automatically disposed when the host is shutdown?

A test that verifies that PEM keys are cleaned up on shutdown would be a good idea.

bartonjs replied:
> A test that verifies that PEM keys are cleaned up on shutdown would be a good idea.

Well.  Sort of.  That's a really hard test to write.  It requires plumbing in a lot of knowledge of Windows Cryptography that even the crypto layer doesn't use automated tests for, since it's potentially subject to change.

davidfowl replied:
> aspnet people, is the certificate automatically disposed when the host is shutdown?

No it's not. We don't call dispose anywhere on the cert provided to Kestrel AFAIK

Tratcher replied:
But it would get finalized.

bartonjs replied:
.NET Core doesn't guarantee finalizers on exit (IIRC).  Unless the host (Kestrel?) calls GC.Collect(); GC.WaitForPendingFinalizers(); on a normal shutdown.

bartonjs replied:
To be fair, the behavior of having used this method is the same as whatever happens when someone specifies a PFX.  Or loads a PFX themselves and passes in an object (if that's a thing).  Since, well, it's just loading a PFX :smile:.

JamesNK replied:
If an app is shutting down (e.g. ctrl+c was pressed), would the process wait around to run finalizers?

I think we should track "temporarily persisted" X509 certs and ensure they are disposed.

javiercn replied:
> To be fair, the behavior of having used this method is the same as whatever happens when someone specifies a PFX. Or loads a PFX themselves and passes in an object (if that's a thing). Since, well, it's just loading a PFX 😄.

Yep this was my reasoning in doing so.

javiercn replied:
> I think we should track "temporarily persisted" X509 certs and ensure they are disposed.

We haven't done this since we added support for config (we should track regular pfx files too if that's the case) and this hasn't been an issue in practice. While I agree that the issue is real, I'm not sure it is something that happens regularly in practice.

I'm not sure if this is because it is hard to reach the upper limit of keys in the underlying abstraction or if it is because people is very diligent in disposing their certificates. I'm siding more with 1. I suggest that we place tracking and disposal of loaded certificates as a separate item.

WRT to loading the keys, the code in here as @bartonjs mentions is equivalent to what happens when you create a new `X509Certiificate2` instance by hand from a pfx.

javiercn replied:
I've filed https://github.com/dotnet/aspnetcore/issues/23622 for tracking disposing the certs on graceful shutdown.

JamesNK commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+                key = rsaKey;
+                return true;
+            }
+            catch
```
try/catch is a little yuck. Is this a common pattern when loading keys?

bartonjs replied:
In the crypto layer we don't guess in new API, you have to be right.  You could use the PemEncoding.TryFind API to read the PEM header and call the correct API for the file type you noticed.

javiercn replied:
> try/catch is a little yuck. Is this a common pattern when loading keys?

I figured out I can probably just look at the public key on the cert to make the right call.

blowdart commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
-                var env = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
-                return new X509Certificate2(Path.Combine(env.ContentRootPath, certInfo.Path), certInfo.Password);
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                if (certInfo.KeyPath != null)
```
Can we not infer from file extension here? Seems a bit weird to move pfx till last if we know how to handle pfx.

javiercn replied:
We don't infer from the file extension here.

If you give us Path, only we assume it is a pfx. If you give us Path and KeyPath we assume it is a cert + key in PEM format.

blowdart commented:
How long will ECC support take to add?

bartonjs commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                if (certInfo.KeyPath != null)
+                {
+                    X509Certificate2 certificate = null;
+                    if (TryReadPemRSAKey(certInfo, environment, out var rsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate =  publicCertificate.CopyWithPrivateKey(rsaKey);
+                    }
+
+                    if (TryReadPemDSAKey(certInfo, environment, out var dsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate = publicCertificate.CopyWithPrivateKey(dsaKey);
+                    }
+
+                    if (certificate != null)
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            return PersistKey(certificate);
+                        }
+
+                        return certificate;
+                    }
```
All of this is logically just

```C#
X509Certificate2 certificate = X509Certificate2.CreateFromPemFile(
    Path.Combine(environment.ContentRootPath, certInfo.Path),
    Path.Combine(environment.ContentRootPath, certInfo.KeyPath));
```

using the new net5 API (checked in already, might not be in a released preview)

If that throws you could fall back to `CreateFromEncryptedPemFile(certPath, certInfo.Password, keyPath)`, assuming that's a scenario.

Alternatively, using the API that already exists, you can use cert.PublicKey.Oid.Value to see what algorithm the cert's public key is, which would tell you what kind of private key to expect.  The problem with that approach is that you need to update here any time new algorithms come along.

bartonjs replied:
I see that the current tests are using a DER-encoded cert along with a PEM-encoded key. That's not a scenario that we support at the lower level, we expect PFX (existing routines) or PEM+PEM (the new methods).

The approach here with CopyWithPrivateKey will get exceptions if the cert file is a PFX and already had a private key, then you loaded a private key.

I'd recommend only supporting PEM+PEM (the new method).  Extracting the signing cert from an authenticode-signed executable and mating it with a PEM-based key is a weird scenario to allow.

javiercn replied:
> The approach here with CopyWithPrivateKey will get exceptions if the cert file is a PFX and already had a private key, then you loaded a private key.

Wouldn't that fail to load because we wouldn't even be passing a password to it?

> I see that the current tests are using a DER-encoded cert along with a PEM-encoded key. That's not a scenario that we support at the lower level, we expect PFX (existing routines) or PEM+PEM (the new methods).

Will this be a problem as long as we check that the certificate loaded in advance doesn't have an associated private key or are there any other issues we should be aware of.

I want to avoid forcing people to convert their certificate format to do this and this is something that I've seen used in other web frameworks too. (DER Cert + PEM key), so I would like that combo to work out of the box, provided that the cert doesn't already contain a private key.

bartonjs replied:
>>     The approach here with CopyWithPrivateKey will get exceptions if the cert file is a PFX and already had a private key, then you loaded a private key.

> Wouldn't that fail to load because we wouldn't even be passing a password to it?

If it had a NULL password or an empty password it will load using the simple ctor.  If it's a PFX with a password other than those two values, then it'd throw an exception for the password being wrong.

> I want to avoid forcing people to convert their certificate format to do this

In my experience (and certainly in the Linux+Let's Encrypt+certbot case) a PEM key is paired with a PEM cert; particularly because they'll use openssl commands, which use PEM output unless you tell it otherwise.

> this is something that I've seen used in other web frameworks too. (DER Cert + PEM key)

Can you give some examples?

> Will this be a problem as long as we check that the certificate loaded in advance doesn't have an associated private key or are there any other issues we should be aware of.

The X509Certificate2 constructor, much to my chagrin, doesn't do simple things.  It does complex things.  The input data can be:

* X.509 public key certificate, binary encoded
* X.509 public key certificate, PEM encoded
* (Not Linux today, bug should be fixed later in 5) A PKCS#7 SignedCms object, binary encoded, where the first signer cert is used
* (Not Linux today, bug should be fixed later in 5) A PKCS#7 SignedCms object, PEM encoded, where the first signer cert is used
* (Windows-only) An authenticode-signed exe, dll, msi, msu, etc, where the first signer cert is used.
* (Windows-only) The SerializedCert format (Windows-proprietary) and probably other things.
* A PKCS#12 PFX, binary encoded

If you really wanted to support both the binary and PEM encoding of a certificate, but restrict the inputs to more sensible things, you could use X509Certificate2.GetCertContentType to ask what it is, and only accept X509.  That still eliminates the case where you load a PFX and have a keyfile specified (and avoids writing the key to disk at all in that error case).

javiercn replied:
> Can you give some examples?

Many dev tools support passing the der cert + the pem keyand handle it transparently. I was doing something similar the other day with the nodejs proxy middleware and they support this transparently. If we can reasonably do it, I don't see why we wouldn't.

I think that with the changes you suggested to how we load the keys it should be enough to support this without problems or keys leaking in error cases.

bartonjs commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+                var rsaKey = RSA.Create();
+                if (certInfo.Password == null)
+                {
+                    rsaKey.ImportFromPem(File.ReadAllText(keyPath));
```
There are currently potentially 6 different calls to File.ReadAllText before the key gets loaded (once you add ECDSA).  Seems like it should be read once, by the caller of this method, then passed along.

Or just change to using CreateFromPemFile.

bartonjs commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+            try
+            {
+                var keyPath = Path.Combine(environment.ContentRootPath, certInfo.KeyPath);
+                var dsa = DSA.Create();
```
If you continue with the manually-loading-keys approach, you should dispose this in your catch block,

davidfowl reviewed (dismissed):
I'd like to discuss this a bit before merging @javiercn to make sure it aligns with the other features we're doing here.

bartonjs commented on src/Servers/Kestrel/shared/test/TestCertificates/https-dsa.key:
```diff
@@ -0,0 +1,23 @@
+-----BEGIN DSA PRIVATE KEY-----
+Proc-Type: 4,ENCRYPTED
+DEK-Info: DES-EDE3-CBC,25877371F267513A
```
We don't support this style of encryption, our PEM support is limited to RFC 7468 PEM, which forbids attributes.

JamesNK replied:
What is the user experience when loading an unsupported key?

We want to give a good error message On startup if the key won’t work

javiercn replied:
Thanks, can I just strip the attributes? I generated these with openssl, I'll try to generate this one without attributes

javiercn replied:
> What is the user experience when loading an unsupported key?

We throw, telling you that either we couldn't find the key or that it is not in a valid format. I have a test to cover that.

bartonjs replied:
> can I just strip the attributes?

No, there are two forms of making an encrypted private key:

1) Build the plaintext key, encrypt it and stuff it into an encrypted key envelope (then use the PEM header type of `-----BEGIN ENCRYPTED PRIVATE KEY-----`)
2) Build the plaintext key, encrypt it and write down that you encrypted it using attributes (keep the PEM header type of `-----BEGIN PRIVATE KEY-----`).

Theoretically you could stuff the payload here into the PKCS#8 EncryptedPrivateKeyInfo format and change the header, but it's a sort of annoying amount of work.

If the openssl CLI isn't doing a friendly job, you can always use a plaintext key, use dsa.ImportFromPem, then dsa.ExportEncryptedPkcs8PrivateKey and write the output as PEM.  ($"-----BEGIN ENCRYPTED PRIVATE KEY-----{Environment.NewLine}{Convert.ToBase64String(data, somethingAboutNewlines)}{Environment.NewLine}-----END ENCRYPTED PRIVATE KEY-----").

javiercn replied:
Yeah, I figured this morning. I have an update that I'm sending where i've created them in pkcs8 format.

javiercn replied:
I've updated all the tests to use pkcs8 keys.

javiercn commented:
> How long will ECC support take to add?

Like 30 minutes? I forgot to add them in the first place

blowdart commented:
Hehe. OK if this gets updated for ECC then I'm good with it, once Jeremy's concerns are addressed

javiercn committed:
Cleanups

javiercn committed:
Remove unnecessary tests

bartonjs commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -480,54 +476,89 @@ static X509Certificate2 PersistKey(X509Certificate2 fullCertificate)
+
+            static X509Certificate2 GetCertificate(CertificateConfig certInfo, string certificatePath, string certificateKeyPath)
+            {
+                if (X509Certificate2.GetCertContentType(certificatePath) != X509ContentType.Unknown)
```
The PEM format certs should come back as X509, not Unknown, meaning this should always go into this if and never hit the later return.

javiercn replied:
I changed it to check explicitly for the Cer type and load the key separately in both cases.

javiercn commented:
* Removed the probing in favor of using the OIDs in the public key of the cert to figure out the private key type.
* Added suppport for PEM cert + PEM Key
* Check the certificate file to determine content type instead of invoking the X509Certificate2 constructor on random data.

### Open questions:
- [ ] I haven't gotten a DSA certificate to load. I've tried all the magic incantations with openssl and even wrote my own, but couldn't make it work.
- [ ] Are there more OIDs we need to probe for?

/cc: @bartonjs

javiercn committed:
Address feedback, add DSA certificate tests

javiercn commented:
🆙📅 

The only thing left is to determine is if we need to handle more OIDs to support the current keys.

javiercn commented:
> I'd like to discuss this a bit before merging @javiercn to make sure it aligns with the other features we're doing here.

We've discussed this offline, I've capture the feedback [here](https://github.com/dotnet/aspnetcore/issues/23623). I think we are still missing some update on the runtime before we can load certificate collection/chains, but that should build on top of this PR.

Once we have an updated BCL with these APIs we can complete the work described in #23623 
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2Collection.cs#L339

I think this is ready to go % build passes

/cc: @davidfowl

JamesNK commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,124 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+            }
+        }
+
+        private static X509Certificate2 TryAttachPemRSAKey(X509Certificate2 certificate, string keyText, string password)
```
The TryAttach methods look like they do the same thing, but with different algorithms.

You could remove some duplication here by resolving an `AsymmetricAlgorithm` from the oid, and then having a single method that loads the cert using the given algorithm.

halter73 replied:
Nit: "Try" implies a boolean return value. Something like GetCertWithAttachedKeyOrNull is a little better.

javiercn replied:
I've refactor the method to make the decision in the caller and just handle the loading here.

halter73 commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,124 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+
+            static X509Certificate2 GetCertificate(CertificateConfig certInfo, string certificatePath, string certificateKeyPath)
+            {
+                if (X509Certificate2.GetCertContentType(certificatePath) == X509ContentType.Cert)
```
Can we log a specific error when the specified cert is of the wrong type?

javiercn replied:
Done

halter73 commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -435,14 +436,124 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+
+                    if (certificate != null)
+                    {
+                        certificate = LoadCertificateKey(certificate, certificateKeyPath, certInfo.Password);
```
It looks like LoadCertificateKey ever return null if the original cert doesn't have one of the 3 expected OIDs. We should log if this happens.

It also looks like the following call to PersistKey will nullref, so we should probably short-circuit in that case.

javiercn replied:
You are totally right, this was an oversight on my part. I've fixed it and added logging to capture the scenarios in which a cert can be null for some reason.

javiercn committed:
Final feedback and cleanups

javiercn committed:
Update from warning to error

davidfowl commented:
@javiercn a couple follow up items:
- Certificate disposal (I see there’s an issue)
- Documentation
- Do we need a public API somewhere that can loads the pem cert and any key in the BCL?

javiercn commented:
> * Do we need a public API somewhere that can loads the pem cert and any key in the BCL?

There's LoadFromPemFile on che BCL, which is waht you would use.

davidfowl commented:
Why don’t we use it?

javiercn commented:
> Why don’t we use it?

Cause the API is PEM + PEM and I wanted to also support loading the cert in DER format, so we load the CERT (PEM or DER) and then we load the key separately and marry it to the cert.

> * Documentation

https://github.com/dotnet/AspNetCore.Docs/issues/19113

> * Certificate disposal (I see there’s an issue)

I would say if it is safe to double dispose certificates we just dispose all certificates on shutdown, otherwise we just dispose the ones loaded from config.

davidfowl commented:
> Cause the API is PEM + PEM and I wanted to also support loading the cert in DER format, so we load the CERT (PEM or DER) and then we load the key separately and marry it to the cert.

Right, should we have an API for this? Or is it too niche?

javiercn commented:
> Right, should we have an API for this? Or is it too niche?

I'm not sure how niche it is, PEM+PEM is likely way more popular, but I want to support if for convenience, so you don't have to use a tool to change the cert format. Some of the dev-servers that I used support this and it is convenient not to require a tool to change the format.

For example, you can use our dev cert with the angular dev server proxy by passing the cert in DER format and the key in PEM format, which is what we plan to do.

bartonjs commented:
> I would say if it is safe to double dispose certificates we just dispose all certificates on shutdown

Yep, certificate disposal is well-behaved.

halter73 commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -429,20 +430,133 @@ private bool TryGetCertificatePath(out string path)
+                        logger.FailedToLoadCertificateKey(certificateKeyPath);
+                    }
+
+                    throw new InvalidOperationException(CoreStrings.InvalidPemKey);
```
This could throw if the cert is missing, not just because the key was missing or invalid right? This exception message implies the key must be the problem?

And why do we use the exact same exception and log message any time the key is missing or invalid? It would be a lot better to log exactly why the key is invalid and to be very clear when the key is actually missing vs being invalid in both the exception and log messages.
